pkgs/build-support/docker/default.nix

```diff
@@ -886,4 +888,70 @@ rec {
         done;
       '' + extraCommands;
     });
+
+  # This function builds a docker image that behaves like a nix-shell
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name
+    , drv
+    , tag ? null
+    # Extra packages available inside the image
+    , extraPackages ? []
+    # Extra bash commands to execute when the image is ran
+    , extraInit ? ""
+    }:
+    # Remove once https://github.com/NixOS/nixpkgs/pull/77952 is merged
+    let
+
+      finalDrv = drv.overrideAttrs (old: {
+        nativeBuildInputs = old.nativeBuildInputs or [] ++ extraPackages;
+      });
+
+      setupPackage = runCommandNoCC "setup" {
+        profile = ''
+          # This mimics part of nix-shell's initialization
+          source $stdenv/setup
+          set +e
+          eval "$shellHook"
+          cd /home
+        '' + extraInit;
+        passAsFile = [ "profile" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/{build,outputs}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
```

Comment:
Why are you not adding `bashInteractive`, `profile`, `coreutils` and `cacert` in the `contents` instead?

Reply:
I want to only include what is absolutely needed. And that's the things added in this derivation, `/bin/sh` and `/usr/bin/env` for POSIX compatibility and the cacert thing to have SSL connectivity. All other files of these derivations aren't needed. And I put the `profile` in here too because it needs to be created in a derivation anyways, so why not use a single one for everything (instead of creating one for the symlinks and another for the profile to be put in `contents`).